File: DataStructure/Set.js
class MySet {
  constructor(k) {
    this.arr = Array.from({ length: k }, (_, i) => i);
  }

  union(x, y) {
    const rootA = this.find(x);
    const rootB = this.find(y);

    this.arr[rootB] = rootA;
  }

  find(x) {
    return (function _find(parents, x) {
      if (parents[x] === x) return x;

      parents[x] = _find(parents, parents[x]); // 경로 압축
      return parents[x];
    })(this.arr, x);
  }

  getNumOfSet() {
    return new Set(this.arr).size;
  }

  getSet() {
    return this.arr;
  }
}

function test(k, operations) {
  const set = new MySet(k);

  for (const [op, val1, val2] of operations) {
    if (op === "u") {
      set.union(val1, val2);
    } else if (op === "f") {
      set.find(val1);
    }
  }

  return set.getNumOfSet();
}

console.log(
  test(3, [
    ["u", 0, 1],
    ["u", 1, 2],
    ["f", 2],
  ])
); //1
console.log(
  test(4, [
    ["u", 0, 1],
    ["u", 2, 3],
    ["f", 3],
  ])
); //2

File: Python/Set.py
class Set() :
    def __init__(self, N) : 
        self.parent = [0] * N
        for i in range(0, N) :
            self.parent[i] = i
        
    def find(self, x) :        
        if self.parent[x] != x :
            return self.find(self.parent[x])
        return x
        
    def union(self, a, b) :
        a = self.find(a)
        b = self.find(b)
        
        if a < b :
            self.parent[b] = a
        else : 
            self.parent[a] = b
            
        print(self.parent)
    
s = Set(11)

s.union(1, 1)
s.union(1, 2)
s.union(1, 3)
s.union(1, 4)
s.union(1, 5)
s.union(2, 6)
s.union(2, 7)
s.union(8, 9)
s.union(8, 10)

print(s.find(10))